File: z_lc/20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        pairs = {'(': ')', '{': '}', '[': ']'}
        stack = []
        for c in s:
            if c in pairs:
                stack.append(pairs[c])
            else:
                if not stack or stack.pop() != c:
                    return False
        return not stack

File: z_lc/20-valid-parentheses/valid-parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        use std::collections::HashMap;
        if s.len() % 2 != 0 { return false; }
        let pairs: HashMap<char, char> = [
            ('(', ')'),
            ('{', '}'),
            ('[', ']'),
        ].iter().cloned().collect();
        let mut stack: Vec<char> = vec![];
        for c in s.chars() {
            match c {
                x if pairs.contains_key(&x) => stack.push(*pairs.get(&x).unwrap()),
                _ => if stack.is_empty() || stack.pop() != Some(c) { return false; },
            }
        }
        stack.is_empty()
    }
}